### DISCUSSION TITLE
ESP32, berry, lvgl and the menu widget

### DISCUSSION THREAD

**Discussion by Cynar** [arendst/Tasmota#22184] [Q&A] [answered]
Hi, I've been trying to wrap my head around doing graphics via tasmota. I'm mostly got my head around the basics, unfortunately I've hit a brick wall with menus. According to the LVGL Berry API Reference, menu is included in tasmota. I cant seem to get it to work.

the most basic initialisation seems to fail.
```
scr = lv.scr_act()
var main_menu=lv.menu(scr)
```
gives:
BRY: Exception> 'attribute_error' - module 'lv' has no member 'menu'
stack traceback:
	<unknown source>: in function `main`

Can anyone point out what I'm missing? I've even tried going through the pyton script that generates the functions, and I cant dig out the menu stuff.

**Comment by s-hadinger** [comment, marked as answer]
LVGL `menu` is not included in precompiled firmwares, because it was considered a feature rarely useful on small screens.

To enable them, you need to add `#define BE_LV_WIDGET_MENU`.

Also I realize it is missing in the commented options in `my_user_config.h`, I will add it.

**Comment by Cynar** [comment]
Thanks for the quick response. I've been trying to compile from source, but I am running into another issue. It seems that be_ntv_lv_menu_init is defined multiple times, and so causes the build to fail.

 /* `lv_menu` methods */
#ifdef BE_LV_WIDGET_MENU
  int be_ntv_lv_menu_init(bvm *vm)       { return be_call_c_func(vm, (void*) &lv_menu_cont_create, "+_p", "(lv.obj)"); }
#endif // BE_LV_WIDGET_MENU
#ifdef BE_LV_WIDGET_MENU
  int be_ntv_lv_menu_init(bvm *vm)       { return be_call_c_func(vm, (void*) &lv_menu_create, "+_p", "(lv.obj)"); }
#endif // BE_LV_WIDGET_MENU
#ifdef BE_LV_WIDGET_MENU
  int be_ntv_lv_menu_init(bvm *vm)       { return be_call_c_func(vm, (void*) &lv_menu_page_create, "+_p", "(lv.obj)c"); }
#endif // BE_LV_WIDGET_MENU
#ifdef BE_LV_WIDGET_MENU
  int be_ntv_lv_menu_init(bvm *vm)       { return be_call_c_func(vm, (void*) &lv_menu_section_create, "+_p", "(lv.obj)"); }
#endif // BE_LV_WIDGET_MENU
#ifdef BE_LV_WIDGET_MENU
  int be_ntv_lv_menu_init(bvm *vm)       { return be_call_c_func(vm, (void*) &lv_menu_separator_create, "+_p", "(lv.obj)"); }
#endif // BE_LV_WIDGET_MENU

in file :
/lib/libesp32_lvgl/lv_binding_berry/generate/be_lv_c_mapping.h

I get the error
`In file included from lib/libesp32_lvgl/lv_binding_berry/src/lv_berry.c:28:
lib/libesp32_lvgl/lv_binding_berry/src/../generate/be_lv_c_mapping.h:1642:7: error: redefinition of 'be_ntv_lv_menu_init'`

On a semi related note. When I use TasmoCompiler, it doesn't seem to include berry properly. The code compiles (when i remove mention of menu) but the button to the berry console is missing. It also doesnt seem to run the autoexec.be file. I'm not sure if it's an issue with TasmoCompiler, me, or something weird in the tasmota base code. I'm assuming i'm being an idiot somewhere, but cant seem to find the issue.

**Reply by s-hadinger** [reply]
Indeed, there is a bug in the mapping script. I will look at fixing this

**Reply by s-hadinger** [reply]
It compiles with the latest commit. Please try it now.
